[AN] refactor: 스크립트 애니메이션 리팩터링

rosemin928 opened the pull request:
<!-- PR 제목 예시
ex. [BE] 로그인 API 구현
ex. [AN] 페이지 구현 -->

## 📌 변경 내용 & 이유
<!-- ex. 
- 사용자 로그인 기능 구현
- 로그인 실패 시 에러 메시지 반환 처리 추가 -->
- 스크립트 화면 recyclerView를 composeView로 변경하며 애니메이션을 더 자연스럽게 수정
- 사용자가 스크롤중일 때는 자동 포커싱 X / 사용자가 다시 스크립트를 눌렀을 때는 포커싱 재시작

## 📸 스크린샷 (선택)
> UI 변경이 있을 경우 첨부

https://github.com/user-attachments/assets/32fc443e-c3b0-4d76-846e-e6b070c2c0a4


## 🧪 테스트 방법 (선택)
<!-- 코드 리뷰하는 팀원이 돌려볼 수 있도록 작성해주기 
ex. ReservationTest의 XXXX 테스트메서드로 확인 
ex. 어떤화면의 어떤기능에서 확인 -->

## 📢 논의하고 싶은 내용
<!-- 이러이러해서 이부분을 이러이러하게 구현했는데 이방식이 괜찮은지? -->

## 🧩 관련 이슈
<!-- 이슈 태그: #123 -->
<!-- 이슈 닫기: close #123 -->
#709 

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * 스크립트 화면을 Compose로 전환해 자동 중앙 정렬, 부드러운 하이라이트 추적 및 클릭 반응 제공
  * 스크립트 전용 ViewModel 추가로 하이라이트·팔로우·사용자 스크롤 상태 실시간 관리

* **Style**
  * 통합 타이포그래피 스타일 세트 추가 및 스크립트 텍스트 색상 애니메이션 적용

* **Bug Fixes**
  * 팔로우/스크롤 동작 안정화로 의도치 않은 자동 스크롤 감소

* **Chores**
  * 기존 RecyclerView 기반 어댑터/뷰홀더/데이터바인딩 및 레이아웃 정리

* **Tests**
  * 일부 UI 테스트 블록 비활성화

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

rosemin928 committed:
[AN] refactor: 필요한 타이포그래피 추가 #709

rosemin928 committed:
[AN] refactor: recyclerView를 composeView로 변경 #709

rosemin928 committed:
[AN] refactor: composeView 변경에 따른 함수 수정 #709

rosemin928 committed:
[AN] refactor: 필요 없는 파일 제거 #709

rosemin928 committed:
[AN] refactor: 누락된 바인딩 추가 #709

rosemin928 committed:
[AN] refactor: 스크립트 클릭 시 발생하는 리플 효과 제거 #709

rosemin928 committed:
[AN] refactor: 스크립트 하이라이트 팔로우 모드 추가 #709

HamBeomJoon commented on android/app/build.gradle.kts:
```diff
@@ -140,6 +140,8 @@ dependencies {
     implementation(composeBom)
     androidTestImplementation(composeBom)
+    implementation(libs.androidx.runtime.livedata)
+    implementation(libs.androidx.lifecycle.runtime.compose)
```
compose만의 livedata인가요??

rosemin928 replied:
Compose에서 LiveData를 관찰하기 위한 라이브러리입니다!
Compose와 LiveData를 연결해주는 어댑터 역할을 한다고 하네요. 해당 라이브러리가 필요한 이유는 아래 코멘트에 또 달아두겠습니다!

HamBeomJoon commented on android/app/src/main/java/com/onair/hearit/presentation/detail/script/ScriptFragment.kt:
```diff
@@ -201,43 +180,62 @@ class ScriptFragment : Fragment() {
-                                ?.scrollToPositionWithOffset(currentIndex, centerOffset)
-                        }
-                    }
+                    highlightedIdState.value = currentItem?.id
```
hightlightedIndex는 List<Script>안에서 하이라이트된 스크립트 인덱스같은데, highlightedIdState는 어디에 쓰는 건가요??

rosemin928 replied:
id는 아이디를 식별하는데 쓰이고, index는 스크롤의 위치를 계산하는 "위치 기반" 동작을 위해 쓰입니다.
현재는 id를 기준으로 삼고, index를 파생해 쓰고 있습니다. 리스트가 바뀌면 index가 쉽게 틀어질 수 있기 때문이 이 방법을 쓰고 있습니다!

HamBeomJoon commented on android/app/src/main/java/com/onair/hearit/presentation/detail/script/component/Scripts.kt:
```diff
@@ -0,0 +1,136 @@
+        snapshotFlow { listState.layoutInfo.visibleItemsInfo.isNotEmpty() }
+            .filter { it }
+            .first()
+    }
```
컴포저블 함수안에 suspend fun이 있는게 뭔가 어색한거같아요 (Fragment와 xml 안에 suspend 함수가 있는 느낌??)
viewModel을 그대로 사용가능한데, 여기서 일부러 하신건가요?

rosemin928 replied:
현재 suspend 함수들은 @Composable 안에 정의되어 있지만 컴포저블에서 직접 호출하는게 아니라 LaunchedEffect 안에서만 호출되고 있어요!
그레서 금지되는 패턴은 아니라고 알고 있는데 뭔가 어색한 느낌인 것은 맞네요 😅
외부 확장 함수나 top-level 함수로 뺄 수는 있을 것 같은데 지금은 함수 자체가 List 레이아웃에 의존을 많이 해서 컴포저블 함수에 써놓았는데 미플은 다른 방식이 더 좋을 것 같다고 생각하시나요??

그리고 함수가 List 레이아웃에 상당히 많이 의존하고 있어서 해당 로직은 ViewModel로 옮기지 않는 것이 좋다고 판단하기도 했습니다!
미플의 생각을 공유해주세용

HamBeomJoon replied:
```
@Stable
class ScriptScrollState(
    val listState: LazyListState,
    private val scriptLines: () -> List<ScriptLine>,
) {
    var isAuto by mutableStateOf(false)
        private set
    var lastCentered by mutableIntStateOf(-1)
        private set

    suspend fun centerVariableHeight(index: Int) {
        val lines = scriptLines()
        if (index !in lines.indices) return
        
        isAuto = true
        try {
            // 스크롤 로직
            lastCentered = index
        } finally {
            isAuto = false
        }
    }
}

@Composable
fun rememberScriptScrollState(
    scriptLines: () -> List<ScriptLine>,
): ScriptScrollState {
    val listState = rememberLazyListState()
    return remember { ScriptScrollState(listState, scriptLines) }
}

@Composable
fun Scripts(scriptLines: List<ScriptLine>, ...) {
    val scrollState = rememberScriptScrollState { scriptLines }
    
    LaunchedEffect(highlightedIndex, followHighlight) {
        if (followHighlight) {
            scrollState.centerVariableHeight(highlightedIndex)
        }
    }
    
    LazyColumn(state = scrollState.listState) { ... }
}
```

아예 클래스로 캡슐화? 하라고 추천해주긴하는데 이런쪽으로 한번 보시는것도 괜찮아보여요. 근데 지금방식도 뭐 어쩔수없는 것 같긴합니다.

rosemin928 replied:
미플이 말씀해주신대로 state를 클래스로 분리해보았습니다!
이렇게 하니 scripts 컴포저블도 훨씬 깔끔하고 보기도 좋은 것 같네요 👍

[[AN] refactor: 스크립트 스크롤 로직을 ScriptScrollState로 분리](https://github.com/woowacourse-teams/2025-hEARit/pull/710/commits/954b013fca2cc605fa5f6774c8c9909878636ab7)

HamBeomJoon commented on android/app/src/main/java/com/onair/hearit/presentation/detail/script/component/Scripts.kt:
```diff
@@ -0,0 +1,136 @@
+) {
+    val listState = rememberLazyListState()
+
+    var isAuto by remember { mutableStateOf(false) }
```
isAuto는 자동으로 스크립트가 이동하는 상태?인가요

rosemin928 replied:
네넹 스크롤이 사용자가 스크롤인지 자동 스크롤인지 구분하는 플래그로 사용하고 있습니당

HamBeomJoon commented on android/app/src/main/java/com/onair/hearit/presentation/detail/script/component/Scripts.kt:
```diff
@@ -0,0 +1,136 @@
+
+    LazyColumn(
+        state = listState,
+        verticalArrangement = Arrangement.spacedBy(0.dp),
```
컴포즈를 잘 모르는데, 0.dp인데 뭔가 다른 동작을 하나요?

rosemin928 replied:
명시적으로 작성한 것 뿐이고, 다른 동작을 하는건 없어서 불필요한 코드라 판단해 삭제했습니다!

HamBeomJoon commented on android/app/src/main/java/com/onair/hearit/presentation/detail/script/component/Scripts.kt:
```diff
@@ -0,0 +1,136 @@
+        state = listState,
+        verticalArrangement = Arrangement.spacedBy(0.dp),
+        modifier =
+            Modifier
+                .fillMaxSize()
+                .pointerInput(Unit) {
+                    detectDragGestures(
+                        onDragStart = {
+                            onUserScrollStateChange(true)
+                            onStopFollow()
+                        },
+                        onDragEnd = { onUserScrollStateChange(false) },
+                        onDragCancel = { onUserScrollStateChange(false) },
+                    ) { _, _ -> }
+                },
```
```
LaunchedEffect(listState) {
    snapshotFlow { listState.isScrollInProgress }.collect { inProgress ->
        onUserScrollStateChange(inProgress)
        if (inProgress && !isAuto) {
            onStopFollow()
        }
    }
}
```
- 위에서 감지하는 함수랑 별개로 또 감지해야하는건가요??
- 드래그 이벤트를 처리하는 부분이 비어있는데 `{ _, _ -> }`, Drag Gesture를 감지만 하는건가요?

rosemin928 replied:
> 1. 위에서 감지하는 함수랑 별개로 또 감지해야하는건가요??

isScrollInProgress는 전체 스크롤 상태를 관찰해서 자동 스크롤인지 유저 스크롤인지를 구분하고, detectDragGestures는 유저의 드래그 시작과 종료를 **즉각적으로** 감지해서 스크립트 팔로우를 즉시 해제합니다.

기능이 어느정도 겹치기는 하지만 detectGestures를 이용한 해당 코드는 손가락이 닿는 그 순간에 팔로우를 해제해서 사용자가 드래그를 끝내기 전에 스크립트가 움직이는 일이 없도록 하는 것이 중요 기능이라고 생각했습니다!


> 2. 드래그 이벤트를 처리하는 부분이 비어있는데 { _, _ -> }, Drag Gesture를 감지만 하는건가요?

detectDragGestures의 시그니처는 이렇게 되어 있어요! :
```kotlin
detectDragGestures(
    onDragStart: (Offset) -> Unit = {},
    onDragEnd: () -> Unit = {},
    onDragCancel: () -> Unit = {},
    onDrag: (change: PointerInputChange, dragAmount: Offset) -> Unit
)
```
현재는 onDragStart / onDragEnd / onDragCancel만 써서 좌표와 거리 정보는 따로 안 쓰고 있기 때문에 비워두었습니다!
드래그 거리만큼 리스트를 직접 스크롤하는 기능은 쓰지 않고, 실제 스크롤은 여전히 LazyColumn이 하되, detectDragGestures는 **"유저가 드래그 중이다"라는 이벤트만 보도록** 했습니다.

HamBeomJoon commented on android/app/src/main/java/com/onair/hearit/presentation/detail/script/ScriptFragment.kt:
```diff
@@ -201,43 +180,62 @@ class ScriptFragment : Fragment() {
+            ViewCompositionStrategy.DisposeOnViewTreeLifecycleDestroyed,
+        )
+        binding.cvScript.setContent {
+            val hearit = viewModel.hearit.observeAsState().value
+            val scripts: List<ScriptLine> = hearit?.script.orEmpty()
+
```
다른 변수들은 stateFlow를 사용하셨는데, 이건 liveData를 사용하고 있는 거 같아요

rosemin928 replied:
현재 hearit을 playerDetail 화면과 뷰모델을 공유하며 사용하고 있는데, stateFlow는 xml에 바인딩이 불가해서 그대로 liveData로 두었습니다. 그리고 이것이 위의 라이브러리를 사용한 이유이기도 합니다. (Compose에서 liveData를 관찰하기 위해...🥲)

이후에 playerDetail 화면도 컴포즈 마이그레이션을 하게 된다면 해당 부분도 같이 stateFlow로 바꿔보겠습니다!

HamBeomJoon replied:
아  Detail이랑 같이 쓰는변수때문에 그렇다는거죠?? 알겠습니다 😅

HamBeomJoon commented on android/app/src/main/java/com/onair/hearit/presentation/detail/script/ScriptFragment.kt:
```diff
@@ -201,43 +180,62 @@ class ScriptFragment : Fragment() {
+                    mediaController?.seekTo(item.start)
+                },
+                onUserScrollStateChange = { scrolling ->
+                    isUserScrollingState.value = scrolling
```
UI 레이어에서 StateFlow에 값을 직접 할당하면 어떤 문제가 발생할 수 있을까요?

HamBeomJoon commented on android/app/src/main/java/com/onair/hearit/presentation/detail/script/ScriptFragment.kt:
```diff
@@ -201,43 +180,62 @@ class ScriptFragment : Fragment() {
+                onUserScrollStateChange = { scrolling ->
+                    isUserScrollingState.value = scrolling
+                    if (scrolling) {
+                        lastUserScrollTime = System.currentTimeMillis()
```
얘도 viewModel에서 할 수 있을 것 같아요

HamBeomJoon reviewed (changes requested):
고생하셨습니다~ Compose에 저도 익숙하지 않아 궁금한 것들 위주로 커멘트 남겼어요!

rosemin928 committed:
Merge branch 'develop-an' into refactor/#709-script-animation-refactor

rosemin928 committed:
[AN] refactor: 어댑터 패키지 분리 #709

rosemin928 committed:
[AN] feat: 스크립트 상태 관리용 ScriptViewModel 추가 #709

rosemin928 committed:
[AN] refactor: 불필요한 코드 제거 #709

rosemin928 committed:
[AN] refactor: LiveData-Compose 라이브러리 이름 수정 #709

- `libs.versions.toml` 파일의 `androidx-runtime-livedata` 항목을 `androidx-compose-runtime-livedata`로 변경합니다.
- `app/build.gradle.kts` 파일에서 수정된 라이브러리 이름을 적용합니다.

HamBeomJoon reviewed (commented):
CI 터지는거랑 의견남긴거만 수정해주세요~!

rosemin928 committed:
[AN] fix: 터지는 테스트 주석 처리 #709

rosemin928 committed:
Merge branch 'develop-an' into refactor/#709-script-animation-refactor

rosemin928 committed:
[AN] fix: libs.versions 정리 #709

rosemin928 committed:
[AN] fix: gradle 정리 #709

rosemin928 committed:
[AN] refactor: 스크립트 스크롤 로직을 ScriptScrollState로 분리 #709

HamBeomJoon reviewed (approved):
고생하셨습니다~
